pootle/apps/virtualfolder/models.py

```diff
@@ -0,0 +1,132 @@
+    )
+    filter_rules = models.TextField(
+        # Translators: This is a noun.
+        _('Filter'),
+        blank=False,
+        help_text=_('Filtering rules that tell which stores this virtual '
+                    'folder comprises.'),
+    )
+    priority = models.FloatField(
+        _('Priority'),
+        default=1,
+        help_text=_('Number specifying importance. Greater priority means it '
+                    'is more important.'),
+    )
+    is_browsable = models.BooleanField(
+        _('Is browsable?'),
+        default=False,
+        help_text=_('Whether this virtual folder is active or not.'),
+    )
+    description = MarkupField(
+        _('Description'),
+        blank=True,
+        help_text=_('Use this to provide more information or instructions. '
+                    'Allowed markup: %s', get_markup_filter_name()),
+    )
+    units = models.ManyToManyField(
+        Unit,
+        db_index=True,
+        related_name='vfolders',
+    )
+
+    class Meta:
+        unique_together = ('name', 'location')
+        ordering = ['-priority', 'name']
+
+    @classmethod
+    def get_matching_for(cls, pootle_path):
+        """Return the matching virtual folders in the given pootle path.
+
+        Not all the applicable virtual folders have matching filtering rules.
+        This method further restricts the list of applicable virtual folders to
+        retrieve only those with filtering rules that actually match.
+        """
+        return VirtualFolder.objects.filter(
+            units__store__pootle_path__startswith=pootle_path
+        ).distinct()
+
+    def __unicode__(self):
+        return ": ".join([self.name, self.location])
+
+    def save(self, *args, **kwargs):
+        if not self.priority > 0:
+            raise Exception(u'Priority is not greater than zero')
+
+        self.name = self.name.lower()
+
+        super(VirtualFolder, self).save(*args, **kwargs)
+
+        # Clean any existing relationship between units and this vfolder.
+        self.units.clear()
+
+        # Recreate relationships between this vfolder and units.
+        for location in self.get_all_pootle_paths():
+            for filename in self.filter_rules.split(","):
+                vf_file = "".join([location, filename])
+
+                qs = Store.objects.filter(pootle_path=vf_file)
+
+                if qs.exists():
+                    self.units.add(*qs[0].units.all())
+
+    def get_all_pootle_paths(self):
+        """Return a list with all the locations this virtual folder applies.
+
+        If the virtual folder location has no {LANG} nor {PROJ} placeholders
+        then the list only contains its location. If any of the placeholders is
+        present, then they get expanded to match all the existing languages and
+        projects.
+        """
+        if "{LANG}" in self.location and "{PROJ}" in self.location:
+            locations = []
+            temp = ""
```

Comment:
When you save a vfolder, if its location contains `{LANG}` or `{PROJ}` placeholders they have to be replaced by all the existing language and project codes. This is necessary to have all the locations the vfolder applies to, so they can be concatenated with the file paths and thus retrieve the stores (if they exist) so the vfolder can create relationships with their units.